template/template.go

```diff
@@ -93,47 +94,114 @@ func (e *executer) execute() error {
 	case nil:
 		return nil
 	case ast.SimpleMessage:
-		return e.resolveSimpleMessage(message)
+		for _, pattern := range message {
+			if err := e.resolvePattern(pattern); err != nil {
+				return fmt.Errorf("resolve pattern: %w", err)
+			}
+		}
 	case ast.ComplexMessage:
-		return errors.New("complex message not implemented") // TODO: Implement.
+		return e.resolveComplexMessage(message)
+	}
+
+	return nil
+}
+
+func (e *executer) resolveComplexMessage(message ast.ComplexMessage) error {
+	if err := e.resolveDeclarations(message.Declarations); err != nil {
+		return fmt.Errorf("resolve declarations: %w", err)
+	}
+
+	if err := e.resolveComplexBody(message.ComplexBody); err != nil {
+		return fmt.Errorf("resolve complex body: %w", err)
 	}
+
+	return nil
 }
 
-func (e *executer) resolveSimpleMessage(message ast.SimpleMessage) error {
-	for _, pattern := range message {
-		switch pattern := pattern.(type) {
-		case ast.TextPattern:
-			if err := e.write(string(pattern)); err != nil {
-				return err
+func (e *executer) resolveDeclarations(declarations []ast.Declaration) error {
```

Comment:
please add duplicate declaration error handling, and test for it
https://github.com/unicode-org/message-format-wg/blob/122e64c2482b54b6eff4563120915e0f86de8e4d/spec/errors.md#duplicate-declaration